common/util/tree_operations.h

```diff
@@ -0,0 +1,926 @@
+template <class Node,  //
+          typename Children_ =
+              tree_operations_internal::TreeNodeChildrenTraits<Node>>
+struct TreeNodeTraits : tree_operations_internal::FeatureTraits {
+  using Parent =
+      detected_or_t<tree_operations_internal::UnavailableFeatureTraits,
+                    tree_operations_internal::TreeNodeParentTraits, Node>;
+  using Value =
+      detected_or_t<tree_operations_internal::UnavailableFeatureTraits,
+                    tree_operations_internal::TreeNodeValueTraits, Node>;
+  using Children = Children_;
+};
+
+// Functions operating on tree nodes:
+
+// Returns true when `node` is a leaf, false otherwise.
+template <class T,  //
+          std::enable_if_t<TreeNodeTraits<T>::available>* = nullptr>
+bool is_leaf(const T& node) {
+  return node.Children().empty();
+}
+
+// Descend through children using indices specified by iterator range.
+// Iterator type must dereference to an integral value.
+// This works on any internal node, not just the root.
+// This implementation fatally exits if any indices are out-of-range.
+// The const-ness of the returned reference matches the const-ness of the node
+// argumnent.
+template <class T, class Iterator,  //
+          std::enable_if_t<TreeNodeTraits<T>::available>* = nullptr>
+T& DescendPath(T& node, Iterator start, Iterator end) {
+  auto* current_node = &node;
+  for (auto iter = start; iter != end; ++iter) {
+    auto& children = current_node->Children();
+    const auto index = *iter;
+    CHECK_GE(index, 0);
+    CHECK_LT(index, std::size(children));
+    current_node = &*std::next(children.begin(), index);  // descend
+  }
+  return *current_node;
+}
+
+// Returns the node reached by descending through *(Children().begin()).
+// The const-ness of the returned reference matches the const-ness of the node
+// argumnent.
+template <class T,  //
+          std::enable_if_t<TreeNodeTraits<T>::available>* = nullptr>
+T& LeftmostDescendant(T& node) {
+  T* leaf = &node;
+  while (!leaf->Children().empty()) {
+    leaf = &*leaf->Children().begin();
+  }
+  return *leaf;
+}
+
+// Returns the node reached by descending through Children().back().
+// The const-ness of the returned reference matches the const-ness of the node
+// argumnent.
+template <class T,  //
+          std::enable_if_t<TreeNodeTraits<T>::available>* = nullptr,
+          std::void_t<decltype(std::declval<T>().Children().back())>* = nullptr>
+T& RightmostDescendant(T& node) {
+  T* leaf = &node;
+  while (!leaf->Children().empty()) {
+    leaf = &leaf->Children().back();
+  }
+  return *leaf;
+}
+
+// std::function type representing a printer function in PrintTree.
+template <class Node>
+using PrintTreePrinterFunction = std::function<std::ostream&(
+    std::ostream&, typename TreeNodeTraits<Node>::Value::const_reference)>;
+
+// Pretty-print in tree-form.  Value() is enclosed in parens, and the whole
+// node is enclosed in braces.
+// This variant supports a custom value 'printer'.
```

Comment:
// Maybe worth noting where many of these functions do not require 'Parent()'.

Reply:
I did it the other way around - I've added comments to functions that DO require `Parent()` or `Value()`.